### PR TITLE
Confidence calibration

### PR DESCRIPTION
**Pull request by adityajain07**
Experiments for confidence calibration.

### CODE REVIEW

File: research/confidence_calibration/temperature_scaling.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+            images, labels = images.to(device), labels.to(device)
+            logits = model(images)
+            logits_list.append(logits)
+            labels_list.append(labels)
+
+    # Concatenate all the logits and labels
+    logits = torch.cat(logits_list).to(device)
+    labels = torch.cat(labels_list).to(device)
+    del logits_list
+    del labels_list
```
Comment (adityajain07) [inline research/confidence_calibration/temperature_scaling.py:257]: @f-PLT There is basically a two-step process.
Step 1: Collect all labels and corresponding model predictions

File: research/confidence_calibration/temperature_scaling.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    loss_function = nn.CrossEntropyLoss().to(device)
+    optimizer = optim.LBFGS(
+        [temp_scaling.temperature], lr=learning_rate, max_iter=max_optim_iterations
+    )
+
+    def closure():
+        optimizer.zero_grad()
+        loss = loss_function(temp_scaling(logits), labels)
+        loss.backward()
+        return loss
```
Comment (adityajain07) [inline research/confidence_calibration/temperature_scaling.py:270]: Step 2: Use all of the collected labels and predictions for optimization.

File: research/confidence_calibration/temperature_scaling.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+    # Concatenate all the logits and labels
+    logits = torch.cat(logits_list).to(device)
+    labels = torch.cat(labels_list).to(device)
+    del logits_list
+    del labels_list
+    gc.collect()
+
+    temp_scaling = TemperatureScaling(initial_t=initial_t).to(device)
+    loss_function = nn.CrossEntropyLoss().to(device)
```
Comment (f-PLT) [inline research/confidence_calibration/temperature_scaling.py:261]: Alright, so you are not setting the `reduction` arguments, which is set to `mean` by default.

Unless there is some counter-indication (which could be possible, I don't have enough knowledge about this), would you then be able to use `sum` instead, iterate on smaller portions of your dataset (or even item by item) and then calculate the mean manually at the end?

**Comment by adityajain07** [comment]
@f-PLT can we merge this PR?

**Review by f-PLT — APPROVED** [review]
